File: lib/go/rpc/ServiceStub.go
package rpc

import "time"

type ServiceStub struct {
	*RpcChannel
	maxMsgSeqId      int64
	msgQueue         map[int64]*RpcMsg
	RpcMsgDispatchCB func(msg *RpcMsg) error
}

func (this *ServiceStub) Invoke(msg *RpcMsg) int {
	msg.RequestMsg.MsgSeqId = this.maxMsgSeqId + 1
	msg.RequestMsg.Header.Version = 0
	msg.Time = time.Now().UTC()
	msg.RequestMsg.Header.MsgType = MPT_Bin
	msg.RequestMsg.Header.ServiceName = msg.ServiceName

	ret := this.GetTransport().sendRequestMsg(&msg.RequestMsg)
	if ret != nil {
		return -1
	}
	this.msgQueue[msg.RequestMsg.MsgSeqId] = msg
	return 0
}

func NewServiceStub(rpcChan *RpcChannel) *ServiceStub {
	stub := &ServiceStub{RpcChannel: rpcChan, maxMsgSeqId: 0, msgQueue: make(map[int64]*RpcMsg)}
	return stub
}

func (this *ServiceStub) stubMsgCallBack(respMsg *ResponseMsg) error {
	rpcMsg, ok := this.msgQueue[respMsg.MsgSeqId]
	if !ok {
		return nil
	}
	rpcMsg.ResponseMsg=*respMsg
	this.RpcMsgDispatchCB(rpcMsg)
	delete(this.msgQueue, respMsg.MsgSeqId)
	return nil
}

File: lib/ts/rpc/ServiceStub.ts
import {RpcChannel} from  "./RpcChannle"
import {ResponseMsg} from  "./ResponseMsg"
import {RpcMsg} from  "./RpcMessage"
import {MsgProtocolType} from  "./MsgProtocolType"
import {Int64} from  "./BigInt"
import { Connection } from "./Connection";

export abstract class ServiceStub{
	public msgQueue:Map<Int64,RpcMsg>=new Map<Int64,RpcMsg>()
	maxMsgSeqId=0;
	protected chan:Connection
	constructor(chan:Connection){
		this.chan=chan
	}

	public stubMsgCallBack(respMsg:ResponseMsg):boolean{
		let rpcMsg=this.msgQueue.get(respMsg.msgSeqId)
		if (rpcMsg==undefined){
			return false
		}
		rpcMsg.responseMsg=respMsg
		this.dispatch(rpcMsg)
		this.msgQueue.delete(respMsg.msgSeqId)
		return true
	}

	protected abstract dispatch( msg:RpcMsg):boolean

	protected invoke(msg:RpcMsg):boolean{
		msg.time= new Date().getTime()
		msg.requestMsg.header.version=0;
		msg.requestMsg.header.msgType=MsgProtocolType.MPT_Bin;
		msg.requestMsg.header.serviceName=msg.serviceName;
		let ret= this.chan.getTransport().sendRequestMsg(msg.requestMsg)
		if (!ret)
		{
			console.log('ServiceStub invoke sendRequestMsg error')
		}

		this.msgQueue.set(msg.requestMsg.msgSeqId,msg)
		return true
	};
}